### PR TITLE
API resource refactor

### LINKED ISSUE #3072: Using 'id' consistently in API responses
`id` and `pk` are used seemingly interchangeably in responses from the REST API , which is kind of confusing.

There seems to be a preference in the JS code for `id`, as when a response uses `pk`, it is often renamed to `id`. `pk` is also hints at the underlying database, while `id` is a little more abstract.

### PR DESCRIPTION
**Pull request by rtibbles**
### Summary
Attempts to simplify the API Resource machinery by getting rid of the use of class inheritance in favour of initialization parameters.
Fixes a few reported issues with the API Resource layer:
* No longer relies solely on the existence of an id for whether a model exists on the server already or not, a new flag is now set on newly created models, so that they will be POSTed rather than PATCHed.
* Instead of resolving the attributes of the model directly, instead resolves a (cheaply computed) deepClone of the attributes to prevent the model from becoming reactive if set onto vuex state.
* Moves getParams for models into the getModel method and uses them to specify caching keys, meaning that models with the same id but different getParams will be cached separately.
* Due to the above change, removed the ability to pass in extra getParams to the `fetch` method.

This PR also adds schema checking in dev mode, to give warnings to developers as they are adding new API Resources regarding the available API Resources on the server, and the methods those resources make available.

This PR also adds a range of helper functions to the Resource class to allow methods to be very simply written for accessing additional endpoints on a DRF viewset. See the TaskResource and ContentNodeSlimResource for details.

This PR further migrates all the content node recommendation filters to be separate endpoints on the content node slim endpoint, utilizing the utility functions above.

Finally, it changes all our returned ids to `id` not `pk`.

### Reviewer guidance
I have manually tested and not found anything broken yet, but it is possible I have missed things.

I would like some feedback on what I have done, and whether it makes sense. Some of the work could probably use some additional automated testing, but most of the refactor is covered by existing tests.

Finally, one thing I have not done is to alter the redundant `getCollection().fetch()` idiom, which could be more succinctly done as `fetchCollection()`. The only reason I have not done this is because it is a major change to the API, but should be relatively trivial to implement. I will happily do so here based on feedback.

### References
Fixes #3072 

…

----

### Contributor Checklist

- [ ] Contributor has fully tested the PR manually
- [ ] PR has the correct target branch and milestone
- [x] PR has 'needs review' or 'work-in-progress' label
- [ ] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label

### Reviewer Checklist

- [ ] Automated test coverage is satisfactory
- [ ] Reviewer has fully tested the PR manually
- [ ] PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- [ ] External dependencies files were updated (`yarn` and `pip`)
- [ ] Documentation is updated
- [ ] Link to diff of internal dependency change is included
- [ ] CHANGELOG.rst is updated for high-level changes
- [ ] Contributor is in AUTHORS.rst

### CODE REVIEW

File: kolibri/core/assets/src/api-resources/channel.js [hunk 1]
```diff
@@ -5,10 +5,8 @@ import { Resource } from '../api-resource';
  * ChannelResource.getModel(channel_id).delete()
  *
  * @example Only get the channels that are "available" (i.e. with resources on device)
- * ChannelResource.getCollection().fetch({ available: true })
+ * ChannelResource.getCollection({ available: true }).fetch()
```
Comment (indirectlylit) [inline kolibri/core/assets/src/api-resources/channel.js:8]: yes!

File: kolibri/core/assets/src/api-resources/contentNode.js [hunk 2]
```diff
@@ -1,125 +1,17 @@
...
-    }
-    return promise;
-  }
-
-  getCopiesCount(getParams = {}) {
-    const collection = this.createCollection({}, getParams, []);
-    collection.url = (...args) => this.urls[`${this.name}-copies-count`](...args);
-    return collection;
-  }
-
```
Comment (indirectlylit) [inline kolibri/core/assets/src/api-resources/contentNode.js:81]: yes!

File: kolibri/core/assets/src/api-resource.js [hunk 3]
```diff
@@ -552,32 +574,35 @@ export class Resource {
     );
   }
 
+  cacheName(endpointName) {
+    return endpointName ? `endpoint-${endpointName}` : 'default';
+  }
+
+  getCache(type, endpointName) {
```
Comment (indirectlylit) [inline kolibri/core/assets/src/api-resource.js]: A lot of the methods in these classes look internal - could we prefix them with underscores to make clear that they're not part of the public API?

File: kolibri/core/assets/src/api-resource.js [hunk 4]
```diff
@@ -665,169 +669,184 @@ export class Resource {
...
+    return this.getCollection(getParams, detailName, id).fetch();
   }
 
-  unCacheCollection(resourceIds = {}, getParams = {}) {
-    if (!this.hasResourceIds) {
-      if (Object.keys(resourceIds).length && Object.keys(getParams).length) {
-        throw TypeError(
-          `resourceIds and getParams passed to getCollection method of ${this.name} ` +
-            'resource, which does not use resourceIds, only pass getParams for this resource'
+  fetchListCollection(listName, getParams = {}) {
```
Comment (indirectlylit) [inline kolibri/core/assets/src/api-resource.js]: `fetchListModel`, `fetchDetailCollection`, `fetchListCollection` are similarly named.. could use some docs

File: kolibri/core/assets/src/api-resources/index.js [hunk 5]
```diff
@@ -1,96 +1,31 @@
...
+export { default as ExamLogResource } from './examLog';
+export { default as ExamAttemptLogResource } from './examAttemptLog';
+export { default as FacilityDatasetResource } from './facilityDataset';
+export { default as UserProgressResource } from './userProgress';
+export { default as ContentNodeProgressResource } from './contentNodeProgress';
+export { default as DeviceProvisionResource } from './deviceProvision';
+export { default as DevicePermissionsResource } from './devicePermissions';
+export { default as RemoteChannelResource } from './remoteChannel';
+export { default as LessonResource } from './lesson';
+export { default as AttemptLogResource } from './attemptLog';
```
Comment (indirectlylit) [inline kolibri/core/assets/src/api-resources/index.js:31]: weird :)

File: kolibri/core/content/api.py [hunk 6]
```diff
@@ -525,12 +384,150 @@ def ancestors(self, request, **kwargs):
...
+            for node in resume:
+                if node.content_id not in content_ids:
+                    deduped_list.append(node)
+                    content_ids.add(node.content_id)
+
+            queryset = resume.filter(id__in=[node.id for node in deduped_list])
+
+        serializer = self.get_serializer(queryset, many=True)
+        return Response(serializer.data)
+
```
Comment (indirectlylit) [inline kolibri/core/content/api.py:530]: nice new routes to replace the filters

**Review by indirectlylit — COMMENTED** [review]
Looks like some productive and positive changes!

* perhaps we could split `api-resource.js` into three files and call the classes and files `APIResourceModel`, `APIResourceCollection`, and `APIResource` or something more explicit?
* I'm a bit confused about the `set` methods. It seems like it's like `save`, but doesn't actually affect server state? (I know this pre-existed the refactor)

File: kolibri/core/assets/src/api-resource.js (hunk 3 above)
Comment (rtibbles) [reply kolibri/core/assets/src/api-resource.js]: Definitely

**Review by indirectlylit — APPROVED** [review]
approved, pending a few updates as described above

File: kolibri/plugins/coach/assets/src/state/actions/reports.js [hunk 7]
```diff
@@ -401,19 +400,21 @@ export function setReportSorting(store, { sortColumn, sortOrder }) {
...
-        collection_id: classId,
-        last_active_time: threshold,
-      }).fetch();
+      const recentReportsPromise = RecentReportResource.fetchCollection({
+        getParams: {
+          channel_id: channelId,
+          content_node_id: channelData.root,
+          collection_kind: UserScopes.CLASSROOM,
+          collection_id: classId,
+          last_active_time: threshold,
```
Comment (jonboiser) [inline kolibri/plugins/coach/assets/src/state/actions/reports.js:415]: @rtibbles I was getting validation errors sending the long-form date string to the server as a filter param. I made the errors go away it by converting to IsoString. Is that the right fix?

```diff
-          last_active_time: threshold,
+          last_active_time: threshold.toISOString(),
```

Comment (rtibbles) [reply kolibri/plugins/coach/assets/src/state/actions/reports.js:415]: Seems right, before this refactor, the threshold was simply being ignored, so no error was being produced.

**Pull request merged**